File: Python/125.Valid Palindrome.py
#判断一个字符串是否为回文字符串(有效字符为字母和数字，其他的不予考虑,字母不考虑大小写)
class Solution(object):
    def isPalindrome(self, s):
        newS=[i.lower() for i in s if i.isalnum()]
        return newS==newS[::-1]

s = Solution()
print(s.isPalindrome("A man, a plan, a canal: Panama"))

#isalnum 如果是字母或数字返回true，其他返回false

File: C++/String/125.Valid Palindrome.cpp
/*
判断一个字符串是不是回文的。
只考虑字母和数字，忽略空格，标点等其他字符。
*/
/*
思路：使用 isalnum函数判断是否是字母或者数字，字母大小写可以用 toupper函数同一转换成大写。
 */
#include <iostream>
#include <cstring>
#include <ctype.h>
using namespace std;

class Solution {
public:
	bool isPalindrome(string s) {
        for(int i = 0, j = s.size()-1;i < j; ++i,--j)
        {
        	while(i<j && !isalnum(s[i]))	++i;
        	while(i<j && !isalnum(s[j]))	--j;
        	if(toupper(s[i])!=toupper(s[j]))	return false;	
        }
        return true;
    }
};

int main()
{
	Solution s;;
	cout<<s.isPalindrome("A man, a plan, a canal: Panama")<<endl;
	cout<<s.isPalindrome("ab")<<endl;
}

